pkg/codegen/hcl2/model/type_test.go

```diff
@@ -714,3 +715,434 @@ func TestRecursiveObjectType(t *testing.T) {
 	hasOutputs, _ = ContainsEventuals(inputLinkedListType)
 	assert.True(t, hasOutputs)
 }
+
+// Tests that object type annotations can be added and queried correctly.
+func TestObjectTypeAnnotations(t *testing.T) {
+	t.Parallel()
+
+	// Tests that an annotation of a non-interface type (here, *x) can be added and queried correctly, whether added using
+	// NewObjectType, Annotate or append.
+	t.Run("new/x-x", func(t *testing.T) {
+		t.Parallel()
+
+		// Arrange.
+		expected := &x{}
+
+		o := NewObjectType(map[string]Type{}, expected)
+
+		// Act.
+		actual, ok := GetObjectTypeAnnotation[*x](o)
+
+		// Assert.
+		require.True(t, ok, "expected to find an annotation of type *x")
+		require.Equal(t, expected, actual)
+	})
+
+	t.Run("append/x-x", func(t *testing.T) {
+		t.Parallel()
+
+		// Arrange.
+		expected := &x{}
+
+		o := NewObjectType(map[string]Type{})
+		o.Annotations = append(o.Annotations, expected)
+
+		// Act.
+		actual, ok := GetObjectTypeAnnotation[*x](o)
+
+		// Assert.
+		require.True(t, ok, "expected to find an annotation of type *x")
+		require.Equal(t, expected, actual)
+	})
+
+	t.Run("annotate/x-x", func(t *testing.T) {
+		t.Parallel()
+
+		// Arrange.
+		expected := &x{}
+
+		o := NewObjectType(map[string]Type{})
+		o.Annotate(expected)
+
+		// Act.
+		actual, ok := GetObjectTypeAnnotation[*x](o)
+
+		// Assert.
+		require.True(t, ok, "expected to find an annotation of type *x")
+		require.Equal(t, expected, actual)
+	})
+
+	// Tests that when multiple annotations of a non-interface type (here, *x) are added, the first is returned by a
+	// query, whether they are added by NewObjectType, Annotate or append.
+	t.Run("new/x-x-x", func(t *testing.T) {
+		t.Parallel()
+
+		// Arrange.
+		expected := &x{}
+
+		o := NewObjectType(map[string]Type{}, expected, &x{})
+
+		// Act.
+		actual, ok := GetObjectTypeAnnotation[*x](o)
+
+		// Assert.
+		require.True(t, ok, "expected to find the first annotation of type *x")
+		require.Equal(t, expected, actual)
```

Comment:
I think we need pointer equality here instead, same I the 2 cases below.

```suggestion
		require.True(t, expected == actual)
```

Reply:
Great point -- have used `require.Same`, which does just that!